### PR TITLE
feat: correct streetname homonymadditions

### PR DESCRIPTION
**Pull request by pgallik**

### CODE REVIEW

File: src/StreetNameRegistry/Municipality/Events/StreetNameHomonoymAdditionsWereCorrected.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    {
+        public const string EventName = "StreetNameHomonoymAdditionsWereCorrected"; // BE CAREFUL CHANGING THIS!!
+
+        [EventPropertyDescription("Interne GUID van de gemeente aan dewelke de straatnaam is gekoppeld.")]
+        public Guid MunicipalityId { get; }
+
+        [EventPropertyDescription("Objectidentificator van de straatnaam.")]
+        public int PersistentLocalId { get; }
+
+        [EventPropertyDescription("De homonymtoevoegingen in de officiële en (eventuele) faciliteitentaal van de gemeente. Mogelijkheden: Dutch, French, German of English.")]
```
Comment (jvandaal) [inline src/StreetNameRegistry/Municipality/Events/StreetNameHomonoymAdditionsWereCorrected.cs]: homon**ie**mtoevoegingen ipv homon**y**mtoevoegingen

File: src/StreetNameRegistry/Municipality/Commands/CorrectHomonymAdditions.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
+﻿namespace StreetNameRegistry.Municipality.Commands;
+
+using System;
+using System.Collections.Generic;
+using Be.Vlaanderen.Basisregisters.Generators.Guid;
+using Be.Vlaanderen.Basisregisters.GrAr.Provenance;
+using Be.Vlaanderen.Basisregisters.Utilities;
+
+public sealed class CorrectHomonymAdditions : IHasCommandProvenance
```
Comment (jvandaal) [inline src/StreetNameRegistry/Municipality/Commands/CorrectHomonymAdditions.cs]: Can we rename it to CorrectStreetNameHomonymAdditions? Seems to be closer to current naming. Same for command for Removing homonym additions.
image

File: src/StreetNameRegistry/Municipality/MunicipalityStreetName.cs [hunk 3]
```diff
@@ -97,6 +98,50 @@ public void CorrectRetirement(Action<Names, HomonymAdditions, PersistentLocalId>
...
+        {
+            GuardStreetNameStatus(StreetNameStatus.Proposed, StreetNameStatus.Current);
+
+            foreach (var item in homonymAdditions)
+            {
+                if (!HomonymAdditions.HasLanguage(item.Language))
+                {
+                    throw new CannotCorrectNonExistentHomonymAdditionException(item.Language.ToString());
+                }
+            }
```
Comment (jvandaal) [inline src/StreetNameRegistry/Municipality/MunicipalityStreetName.cs:113]: We could change the foreach to if you agree this is more readable?
```
if (homonymAdditions.Any(x => !HomonymAdditions.HasLanguage(x.Language)))
{
    throw new CannotCorrectNonExistentHomonymAdditionException();
}
```

File: src/StreetNameRegistry/Municipality/MunicipalityStreetName.cs [hunk 4]
```diff
@@ -97,6 +98,50 @@ public void CorrectRetirement(Action<Names, HomonymAdditions, PersistentLocalId>
...
+        public void RemoveHomonymAdditions(
+            List<Language> languages,
+            Action<Names, HomonymAdditions, PersistentLocalId> guardUniqueActiveStreetNameNames)
+        {
+            GuardStreetNameStatus(StreetNameStatus.Proposed, StreetNameStatus.Current);
+
+            var names =  new Names(Names.Where(x => languages.Contains(x.Language)));
+            guardUniqueActiveStreetNameNames(names, new HomonymAdditions(), PersistentLocalId);
+
+            var homonymAdditionsToRemove =
```
Comment (jvandaal) [inline src/StreetNameRegistry/Municipality/MunicipalityStreetName.cs]: Can we put this on one line?

File: test/StreetNameRegistry.Tests/AggregateTests/WhenCorrectingHomonymAdditions/GivenHomonymAdditionForStreetNameAlreadyExists.cs [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        Fixture.Customize(new WithFixedMunicipalityId());
+        Fixture.Customize(new WithFixedPersistentLocalId());
+        Fixture.Customize(new InfrastructureCustomization());
+
+        _streamId = Fixture.Create<MunicipalityStreamId>();
+    }
+
+    [Fact]
+    public void GivenAStreetNameWithTheSameHomonymAdditionAlreadyExists_Throws_StreetNameNameAlreadyExistsException()
+    {
```
Comment (jvandaal) [inline test/StreetNameRegistry.Tests/AggregateTests/WhenCorrectingHomonymAdditions/GivenHomonymAdditionForStreetNameAlreadyExists.cs]: Both the class and the method start with `Given`. Can we rename the method to something else? I think it suffices to just leave ThenThrowsStreetNameNameAlreadyExistsException

File: test/StreetNameRegistry.Tests/AggregateTests/WhenCorrectingHomonymAdditions/GivenMissingHomonymAdditionWithLanguage.cs [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+using global::AutoFixture;
+using Municipality;
+using Municipality.Commands;
+using Municipality.Events;
+using Municipality.Exceptions;
+using Testing;
+using Xunit;
+using Xunit.Abstractions;
+
+public class GivenMissingHomonymAdditionWithLanguage : StreetNameRegistryTest
```
Comment (jvandaal) [inline test/StreetNameRegistry.Tests/AggregateTests/WhenCorrectingHomonymAdditions/GivenMissingHomonymAdditionWithLanguage.cs]: I would rename this class to GivenStreetNameWithoutHomonymAddition. They the starting scenario on which you are going to correct a homonym addition is more clear.

File: test/StreetNameRegistry.Tests/AggregateTests/WhenCorrectingHomonymAdditions/GivenMissingHomonymAdditionWithLanguage.cs [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    {
+        Fixture.Customize(new WithFixedMunicipalityId());
+        Fixture.Customize(new WithFixedPersistentLocalId());
+        Fixture.Customize(new InfrastructureCustomization());
+
+        _streamId = Fixture.Create<MunicipalityStreamId>();
+    }
+
+    [Fact]
+    public void ThenThrow_CannotCorrectNonExistentHomonymAdditionException()
```
Comment (jvandaal) [inline test/StreetNameRegistry.Tests/AggregateTests/WhenCorrectingHomonymAdditions/GivenMissingHomonymAdditionWithLanguage.cs]: Rename to ThenThrows or ThrowsCannotCorrectNonExistentHomonymAdditionException.

File: test/StreetNameRegistry.Tests/AggregateTests/WhenCorrectingHomonymAdditions/GivenPassesAllValidations.cs [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+using Be.Vlaanderen.Basisregisters.GrAr.Provenance;
+using global::AutoFixture;
+using Municipality;
+using Municipality.Commands;
+using Municipality.Events;
+using Testing;
+using Xunit;
+using Xunit.Abstractions;
+
+public class GivenPassesAllValidations : StreetNameRegistryTest
```
Comment (jvandaal) [inline test/StreetNameRegistry.Tests/AggregateTests/WhenCorrectingHomonymAdditions/GivenPassesAllValidations.cs]: Unlike the other class names, this class name doesn't give us the functional scenario.

File: test/StreetNameRegistry.Tests/AggregateTests/WhenCorrectingHomonymAdditions/GivenStreetName.cs [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+                Fixture.Create<MunicipalityBecameCurrent>(),
+                streetNameWasMigratedToMunicipality)
+            .When(command)
+            .ThenNone());
+    }
+
+    [Theory]
+    [InlineData(StreetNameStatus.Current)]
+    [InlineData(StreetNameStatus.Proposed)]
+    public void WithValidStatuses_ThenStreetNameHomonymAdditionsCorrectedEvent(StreetNameStatus status)
```
Comment (jvandaal) [inline test/StreetNameRegistry.Tests/AggregateTests/WhenCorrectingHomonymAdditions/GivenStreetName.cs]: WithValidStatuses -> WithValidStatus. A StreetName has only a single status.

File: test/StreetNameRegistry.Tests/AggregateTests/WhenRemovingHomonymAdditions/GivenStreetNameNameWithHomonymAdditionAlreadyExists.cs [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+using global::AutoFixture;
+using Municipality;
+using Municipality.Commands;
+using Municipality.Events;
+using Municipality.Exceptions;
+using Testing;
+using Xunit;
+using Xunit.Abstractions;
+
+public class GivenStreetNameNameWithHomonymAdditionAlreadyExists : StreetNameRegistryTest
```
Comment (jvandaal) [inline test/StreetNameRegistry.Tests/AggregateTests/WhenRemovingHomonymAdditions/GivenStreetNameNameWithHomonymAdditionAlreadyExists.cs:17]: Rename to GivenStreetNameNameWithoutHomonymAdditionAlreadyExists

File: src/StreetNameRegistry/Municipality/MunicipalityStreetName.cs [hunk 11]
```diff
@@ -97,6 +98,47 @@ public void CorrectRetirement(Action<Names, HomonymAdditions, PersistentLocalId>
             Apply(new StreetNameWasCorrectedFromRetiredToCurrent(_municipalityId, PersistentLocalId));
         }
 
+        public void CorrectHomonymAdditions(
+            HomonymAdditions homonymAdditions,
+            Action<Names, HomonymAdditions, PersistentLocalId> guardUniqueActiveStreetNameNames)
```
Comment (ArneD) [inline src/StreetNameRegistry/Municipality/MunicipalityStreetName.cs:103]: new lines

File: src/StreetNameRegistry/Municipality/MunicipalityStreetName.cs [hunk 12]
```diff
@@ -97,6 +98,47 @@ public void CorrectRetirement(Action<Names, HomonymAdditions, PersistentLocalId>
...
+
+            if (changedHomonymAdditions.Any())
+            {
+                Apply(new StreetNameHomonymAdditionsWereCorrected(_municipalityId, PersistentLocalId, changedHomonymAdditions));
+            }
+        }
+
+        public void RemoveHomonymAdditions(
+            List<Language> languages,
+            Action<Names, HomonymAdditions, PersistentLocalId> guardUniqueActiveStreetNameNames)
```
Comment (ArneD) [inline src/StreetNameRegistry/Municipality/MunicipalityStreetName.cs:127]: new lines

File: test/StreetNameRegistry.Tests/ValueObjectTests/StreetNameHomonymAdditionTests.cs [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+    [Fact]
+    public void WhenHomonymAdditionCharacterLimitExceeded_ThenThrowHomonymAdditionMaxCharacterLengthExceededException()
+    {
+        var execute = () => new StreetNameHomonymAddition("homonym exceeding character limit", Language.Dutch);
+
+        execute
+            .Should()
+            .Throw<HomonymAdditionMaxCharacterLengthExceededException>()
+            .Where(item => item.LanguageCode == Language.Dutch.ToString());
+    }
```
Comment (ArneD) [inline test/StreetNameRegistry.Tests/ValueObjectTests/StreetNameHomonymAdditionTests.cs:27]: add `null` check test?

**Review by ArneD — CHANGES_REQUESTED** [review]

File: src/StreetNameRegistry/Municipality/MunicipalityStreetName.cs (hunk 11 above)
Comment (emalfroy) [reply src/StreetNameRegistry/Municipality/MunicipalityStreetName.cs:103]: It is one Action: `Action<Names, HomonymAdditions, PersistentLocalId> guardUniqueActiveStreetNameNames)`

File: src/StreetNameRegistry/Municipality/MunicipalityStreetName.cs (hunk 12 above)
Comment (emalfroy) [reply src/StreetNameRegistry/Municipality/MunicipalityStreetName.cs:127]: It is one Action: `Action<Names, HomonymAdditions, PersistentLocalId> guardUniqueActiveStreetNameNames)`

**Review by ArneD — APPROVED** [review]

**Pull request merged**